pkg/querier/frontend/frontend_test.go

```diff
@@ -251,8 +255,75 @@ func TestFrontendCheckReady(t *testing.T) {
 	}
 }
 
-func testFrontend(t *testing.T, config Config, handler http.Handler, test func(addr string), matchMaxConcurrency bool) {
+func TestFrontend_LogsSlowQueriesFormValues_Issue3111(t *testing.T) {
+	// Create an HTTP server listening locally. This server mocks the downstream
+	// Prometheus API-compatible server.
+	downstreamListen, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	downstreamServer := http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			_, err := w.Write([]byte(responseBody + responseBody + responseBody + responseBody + responseBody))
+			require.NoError(t, err)
+		}),
+	}
+
+	defer downstreamServer.Shutdown(context.Background()) //nolint:errcheck
+	go downstreamServer.Serve(downstreamListen)           //nolint:errcheck
+
+	// Configure the query-frontend with the mocked downstream server.
+	config := defaultFrontendConfig()
+	config.LogQueriesLongerThan = 1 * time.Microsecond
+	config.DownstreamURL = fmt.Sprintf("http://%s", downstreamListen.Addr())
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte(responseBody + responseBody + responseBody + responseBody + responseBody))
+		time.Sleep(10 * time.Millisecond)
+		require.NoError(t, err)
+	})
+	var buf bytes.Buffer
+	l := log.NewLogfmtLogger(log.NewSyncWriter(&buf))
+
+	test := func(addr string) {
+		data := url.Values{}
+		data.Set("test", "form")
+		data.Set("issue", "3111")
+
+		req, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("http://%s/", addr), strings.NewReader(data.Encode()))
+		req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+		req.Header.Add("Content-Length", strconv.Itoa(len(data.Encode())))
+
+		ctx := context.Background()
+		req = req.WithContext(ctx)
+		assert.NoError(t, err)
+		err = user.InjectOrgIDIntoHTTPRequest(user.InjectOrgID(ctx, "1"), req)
+		assert.NoError(t, err)
+
+		client := http.Client{
+			Transport: &nethttp.Transport{},
+		}
+
+		resp, err := client.Do(req)
+		assert.NoError(t, err)
+		b, err := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+		assert.NoError(t, err)
+		assert.Equal(t, 200, resp.StatusCode, string(b))
+
+		logs := buf.String()
```

Comment:
Should we also assert that doesn't contain `invalid Read on closed Body`?

Reply:
Added assertion for this error message